src/axom/slam/RelationSet.hpp

```diff
@@ -29,11 +30,15 @@ namespace slam
  */
 
 template <typename Relation,
-          typename SetType1 = slam::Set<>,
-          typename SetType2 = slam::Set<>>
+          typename SetType1 = typename Relation::FromSetType,
+          typename SetType2 = typename Relation::ToSetType,
+          typename InterfaceType = policies::VirtualInterface>
 class RelationSet final
-  : public OrderedSet<typename Relation::SetPosition, typename Relation::SetElement>,
-    public BivariateSet<SetType1, SetType2>
+  : public policies::BivariateSetInterface<
+      InterfaceType,
+      SetType1,
+      SetType2,
+      RelationSet<Relation, SetType1, SetType2, InterfaceType>>

```

Comment:
OK, maybe I'm just blind, but this looks like recursive templatizing.  RelationSet inherits from BivariateSetInterface whose final template is the same template class RelationSet.  Am I reading this right?  If so, where's the base case, so this doesn't go on recursing?

Reply:
Good question! I'm using the [curiously-recursive template pattern](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) here. It's usually used to achieve similar results to polymorphism without the expense of virtual functions calls.

This will probably go away once I address this https://github.com/LLNL/axom/pull/945#discussion_r1000933708, though.